### PR TITLE
Allow the Realm Gradle plugin to modify the bytecode

### PR DESCRIPTION
**Pull request by emanuelez**
The bytecode is modified in two ways:
 * Custom accessors are added to model classes
    They are called .realm[GS]etter$<fieldName>
 * Such accessors are used instead of field accesses

This is part of making model classes feel like normal POJOs.

### CODE REVIEW

**Comment by emanuelez** [comment]
@cmelchior @nhachicha @kneth

File: gradle-plugin/plugin/build.gradle [hunk 1]
```diff
@@ -1,6 +1,12 @@
 apply plugin: 'groovy'
 
+repositories {
+    jcenter()
+}
+
 dependencies {
     compile gradleApi()
     compile localGroovy()
+    compile 'com.android.tools.build:gradle:1.2.3'
```
Comment (cmelchior) [inline gradle-plugin/plugin/build.gradle]: This should be pulled from the global build.gradle instead

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 2]
```diff
@@ -1,13 +1,126 @@
...
+        // Make sure the project is either an Android application or library
+        def isAndroidApp = project.plugins.withType(AppPlugin)
+        def isAndroidLib = project.plugins.withType(LibraryPlugin)
+        if (!isAndroidApp && !isAndroidLib) {
+            throw new GradleException("'android' or 'android-library' plugin required.")
+        }
+
+        // Configure the project
+        project.repositories.add(project.repositories.jcenter())
         project.dependencies.add('compile', 'io.realm:realm-android:0.80.3') // TODO: make version dynamic
```
Comment (cmelchior) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: Is this incoming?

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 3]
```diff
@@ -1,13 +1,126 @@
...
+        final def variants
+        if (isAndroidApp) {
+            variants = project.android.applicationVariants
+        } else {
+            variants = project.android.libraryVariants
+        }
+
+        // Do the bytecode manipulation
+        variants.all { variant ->
+            variant.dex.doFirst() {// Get the jar files of the imported libraries
```
Comment (cmelchior) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: How does this work with library projects and pre-dexing? Ie. if the app is accessing fields of a Realm class defined in a library project that has been pre-dexed?

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 4]
```diff
@@ -1,13 +1,126 @@
...
+                }
+
+                // Append the folder containing the .class files for the project
+                classPool.appendClassPath(javaCompile.destinationDir.path)
+
+                // Find the class files
+                ConfigurableFileTree classFiles = project.fileTree(javaCompile.destinationDir);
+                classFiles.include("**/*.class");
+
+
```
Comment (cmelchior) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: Remove whitespace or add comment to indicate new "section"

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 5]
```diff
@@ -1,13 +1,126 @@
...
+    }
+
+    private static void useRealmAccessors(CtClass clazz) {
+        clazz.getDeclaredBehaviors().each { behavior ->
+            if (!behavior.name.startsWith('realmGetter$') && !behavior.name.startsWith('realmSetter$')) {
+                behavior.instrument(new ExprEditor() {
+                    @Override
+                    void edit(FieldAccess fieldAccess) throws CannotCompileException {
+                        if (fieldAccess.field?.declaringClass?.superclass?.name?.equals("io.realm.RealmObject")) {
+                            println("Realm: Manipulating ${clazz.simpleName}.${behavior.name}(): ${fieldAccess.fieldName}")
```
Comment (cmelchior) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: Do we really want to flood the log with this?

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 6]
```diff
@@ -1,13 +1,126 @@
...
+                        }
+                    }
+                })
+            }
+        }
+    }
+
+    private static void addRealmAccessors(CtClass clazz) {
+        if (clazz.superclass.name.equals("io.realm.RealmObject")) {
+            println("Realm: Adding accessors to ${clazz.simpleName}")
```
Comment (cmelchior) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: Same

File: gradle-plugin/sample/src/main/java/io/realm/example/sample/models/Person.java [hunk 7]
```diff
@@ -16,10 +16,12 @@
 package io.realm.example.sample.models;
 
 import io.realm.RealmObject;
+import io.realm.annotations.Ignore;
 
 public class Person extends RealmObject {
```
Comment (cmelchior) [inline gradle-plugin/sample/src/main/java/io/realm/example/sample/models/Person.java]: Perhaps also add a setter that mutates the data?
I assume that removing limitations on methods etc. will be another PR?

File: gradle-plugin/plugin/build.gradle (hunk 1 above)
Comment (emanuelez) [inline gradle-plugin/plugin/build.gradle]: Not yet. The plugin is still its own project

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy (hunk 2 above)
Comment (emanuelez) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: I had some discussions with Nabil about this... we will take it all on Thursday cause it's important and has dramatic consequences

**Comment by cmelchior** [comment]
Looks like a great start :dancers: 
Things I would like to see:
- How does this effect build times?
- Does it effect the debugging experience?

I am a bit concerned about the interaction with library projects and Realm classes exposed from there. This probably needs to be tested more.

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy (hunk 3 above)
Comment (emanuelez) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: It's a use-case I thought about and it will require some work. Another topic for Thursday

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy (hunk 5 above)
Comment (emanuelez) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: We do for the annotation processor. Either we do both and none IMO

File: gradle-plugin/sample/src/main/java/io/realm/example/sample/models/Person.java (hunk 7 above)
Comment (emanuelez) [inline gradle-plugin/sample/src/main/java/io/realm/example/sample/models/Person.java]: That will require changes to the annotation processor, which is not part of this PR.
The changes to the annotation processor will have a huge impact on the realm unit tests

**Comment by emanuelez** [comment]
As far as I could see build times are not affected dramatically. We will be able to do proper testing when every piece will be in piece.

The debugging experience should not change since as far as the IDE is concerned those are still field accesses and cannot be stepped into. Again, more testing needed when the changes to the AP are in place.

I see some pretty elegant solutions about the Realm libraries use-case, which we can discuss F2F :)

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 8]
```diff
@@ -1,13 +1,126 @@
...
+        clazz.getDeclaredBehaviors().each { behavior ->
+            if (!behavior.name.startsWith('realmGetter$') && !behavior.name.startsWith('realmSetter$')) {
+                behavior.instrument(new ExprEditor() {
+                    @Override
+                    void edit(FieldAccess fieldAccess) throws CannotCompileException {
+                        if (fieldAccess.field?.declaringClass?.superclass?.name?.equals("io.realm.RealmObject")) {
+                            println("Realm: Manipulating ${clazz.simpleName}.${behavior.name}(): ${fieldAccess.fieldName}")
+                            def fieldName = fieldAccess.fieldName
+                            if (fieldAccess.isReader()) {
+                                fieldAccess.replace('$_ = $0.realmGetter$' + fieldName + '();')
```
Comment (cmelchior) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: We might want to consider replacing $ with _, I think i saw some discussion on this in Dagger repo at one point and some editors (perhaps Eclipse) has problems with this. I will see if I can dig the discussion up.

Comment (emanuelez) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: Ok. It should be fine according to the Java Language Specification: http://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8

File: gradle-plugin/sample/build.gradle [hunk 9]
```diff
@@ -14,11 +14,11 @@ apply plugin: 'realm'
 
 android {
     compileSdkVersion 22
-    buildToolsVersion "21.1.2"
+    buildToolsVersion "22.0.0"
 
     defaultConfig {
         applicationId "io.realm.example.sample"
-        minSdkVersion 15
+        minSdkVersion 22
```
Comment (kneth) [inline gradle-plugin/sample/build.gradle]: We have an issue arguing that we can use version 21.

**Comment by kneth** [comment]
Please post the provide line for `changelog.md`.

**Comment by emanuelez** [comment]
There's no user facing changes since there's no way for a user to apply the plugin yet

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 10]
```diff
@@ -1,13 +1,126 @@
 package io.realm.gradle
-
+import com.android.build.gradle.AppPlugin
+import com.android.build.gradle.LibraryPlugin
+import javassist.*
```
Comment (nhachicha) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: shouldn't be allowed with new checkstyle :)

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy [hunk 11]
```diff
@@ -1,13 +1,126 @@
...
-    	project.repositories.add(project.repositories.jcenter())
+        // Make sure the project is either an Android application or library
+        def isAndroidApp = project.plugins.withType(AppPlugin)
+        def isAndroidLib = project.plugins.withType(LibraryPlugin)
+        if (!isAndroidApp && !isAndroidLib) {
+            throw new GradleException("'android' or 'android-library' plugin required.")
+        }
+
+        // Configure the project
+        project.repositories.add(project.repositories.jcenter())
```
Comment (nhachicha) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: What happen if the user already pulled `jcenter()`?

File: gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy (hunk 5 above)
Comment (nhachicha) [inline gradle-plugin/plugin/src/main/groovy/io/realm/gradle/Realm.groovy]: we can use it with a Debug flag, otherwise it will slow down the gradle task

File: gradle-plugin/sample/build.gradle [hunk 12]
```diff
@@ -33,5 +33,5 @@ android {
 
 dependencies {
     compile fileTree(dir: 'libs', include: ['*.jar'])
-    compile 'com.android.support:appcompat-v7:22.0.0'
+    //compile 'com.android.support:appcompat-v7:22.0.0'
```
Comment (nhachicha) [inline gradle-plugin/sample/build.gradle]: remove dead code

File: gradle-plugin/sample/src/main/java/io/realm/example/sample/models/Person.java [hunk 13]
```diff
@@ -16,10 +16,12 @@
...
     private String name;
     private int age;
+    @Ignore public String note;
```
Comment (nhachicha) [inline gradle-plugin/sample/src/main/java/io/realm/example/sample/models/Person.java]: new checkstyle `@Ignore` on new line

**Comment by nhachicha** [comment]
since this is going to be a long WIP (tests, refactor) maybe we can merge against a feature branch rather than `master`. so you can iterate quickly without worrying to break/regression master

**Comment by zaki50** [comment]
Should we consider using Transform API newly introduced by android gradle plugin 1.4.0-beta2?

http://tools.android.com/tech-docs/new-build-system/transform-api

**Comment by cmelchior** [comment]
Probably yes. The good thing about having a Realm gradle plugin is that we can switch to using the transform API when it becomes available.

**Comment by emanuelez** [comment]
I'm looking into it. It's very promising, but it would force our users to use Gradle 1.4 which might be a bit too much right now

**Comment by dalinaum** [comment]
Should we add the `wip` tag on this?

**Comment by Saketme** [comment]
@emanuelez Do not worry about forcing users to gradle 1.4. It's a one time download thing. And I guess most of us are already on 1.5

**Comment by emanuelez** [comment]
Superseded by #2196
